src/main/java/org/opentripplanner/routing/algorithm/filterchain/filter/SameFirstOrLastTripFilter.java

```diff
@@ -0,0 +1,38 @@
+package org.opentripplanner.routing.algorithm.filterchain.filter;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.opentripplanner.model.SystemNotice;
+import org.opentripplanner.model.plan.Itinerary;
+import org.opentripplanner.routing.algorithm.filterchain.ItineraryListFilter;
+import org.opentripplanner.routing.algorithm.filterchain.groupids.GroupBySameFirstOrLastTrip;
+
+/**
+ * This filter ensures that no more than one itinerary begins or ends with same trip.
+ * It loops through itineraries from top to bottom. If itinerary matches with any other itinerary
+ * from above, it is removed from list.
+ * Uses {@link org.opentripplanner.routing.algorithm.filterchain.groupids.GroupBySameFirstOrLastTrip}.
+ * for matching itineraries.
+ */
+public class SameFirstOrLastTripFilter implements ItineraryListFilter {
+    @Override
+    public List<Itinerary> filter(List<Itinerary> itineraries) {
+        List<GroupBySameFirstOrLastTrip> groups = new ArrayList<>();
+
+        OUTER_LOOP:
+        for (Itinerary it : itineraries) {
+            GroupBySameFirstOrLastTrip currentGroup = new GroupBySameFirstOrLastTrip(it);
+
+            for (GroupBySameFirstOrLastTrip group : groups) {
```

Comment:
Am I seeing this correctly that it's the first one that wins and the rest with the same first/last are discarded?

Presumably for this reason you're also sort by cost so the one with the lowest cost comes first, isn't it?

Reply:
Itinerary with lowest cost should be the one that remains and that will always be the case with we sort beforehand so the "best" result lies on top.